core/src/ast/attrs.rs

```diff
@@ -1,10 +1,201 @@
 //! This module contains utilities for dealing with Rust attributes
 
-use syn::Attribute;
+use serde::ser::{SerializeStruct, Serializer};
+use serde::Serialize;
+use syn::parse::{Error as ParseError, Parse, ParseStream};
+use syn::{Attribute, Ident, LitStr, Meta, Token};
 
-pub(crate) fn extract_cfg_attrs(attrs: &[Attribute]) -> impl Iterator<Item = String> + '_ {
-    attrs
-        .iter()
-        .filter(|&a| a.path().is_ident("cfg"))
-        .map(|a| quote::quote!(#a).to_string())
+/// The list of attributes on a type
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Default)]
+pub struct Attrs {
+    pub cfg: Vec<Attribute>,
+    pub attrs: Vec<DiplomatAttr>,
+}
+
+impl Attrs {
+    fn add_attr(&mut self, attr: Attr) {
+        match attr {
+            Attr::Cfg(attr) => self.cfg.push(attr),
+            Attr::DiplomatAttr(attr) => self.attrs.push(attr),
+        }
+    }
+
+    /// Merge attributes that should be inherited from the parent
+    pub(crate) fn merge_parent_attrs(&mut self, other: &Attrs) {
+        self.cfg.extend(other.cfg.iter().cloned())
+    }
+    pub(crate) fn add_attrs(&mut self, attrs: &[Attribute]) {
+        for attr in syn_attr_to_ast_attr(attrs) {
+            self.add_attr(attr)
+        }
+    }
+    pub(crate) fn from_attrs(attrs: &[Attribute]) -> Self {
+        let mut this = Self::default();
+        this.add_attrs(attrs);
+        this
+    }
+}
+
+impl From<&[Attribute]> for Attrs {
+    fn from(other: &[Attribute]) -> Self {
+        Self::from_attrs(other)
+    }
+}
+
+enum Attr {
+    Cfg(Attribute),
+    DiplomatAttr(DiplomatAttr),
```

Comment:
Suggestion (optional): `DiplomatBackendAttr`

Can also wait for when we add more variants to make such a change